aiida_testing/export_cache/_fixtures.py

```diff
@@ -0,0 +1,375 @@
+                and key not in self._updatable_attributes
+            },
+            #self._repository._get_base_folder(),
+            #self.computer.uuid if self.computer is not None else None
+        ]
+
+        return objects
+
+    # since we still want versioning for plugin datatypes and calcs we only monkeypatch aiida datatypes
+    classes_to_patch = [Dict, SinglefileData, List, FolderData, RemoteData]
+    for classe in classes_to_patch:
+        monkeypatch.setattr(classe, "_get_objects_to_hash", mock_objects_to_hash)
+
+    #BaseData, List, Array, ...
+
+
+@pytest.fixture(scope='function')
+def run_with_cache(export_cache, load_cache):
+    """
+    Fixture to automatically import an aiida graph for a given process builder.
+    """
+    def _run_with_cache(
+        builder: dict,  #aiida process builder class, or dict, if process class is given
+        process_class=None,
+        label: str = '',
+        data_dir: ty.Union[str, pathlib.Path] = 'data_dir',
+    ):
+        """
+        Function, which checks if a aiida export for a given Process builder exists,
+        if it does it imports the aiida graph and runs the builder with caching.
+        If the cache does not exists, it still runs the builder but creates an
+        export afterwards.
+
+        Inputs:
+
+        builder : AiiDA Process builder class,
+        data_dir: optional
+            Absolute path of the directory where the exported workchain graphs are
+            stored.
+
+        ignore_nodes : list string, ignore input nodes with these labels/link labels to ignore in hash.
+        # needed?
+        """
+
+        cache_exists = False
+        if data_dir == 'data_dir':
+            cwd = pathlib.Path(os.getcwd())  # Might be not the best idea.
+            data_dir = (cwd / 'data_dir')  # TODO: get from config?
+
+        bui_hash, input_nodes = get_hash_process(builder)
+
+        if process_class is None:  # and isinstance(builder, dict):
+            process_class = builder.process_class  # type: ignore
+            # we assume ProcessBuilder, since type(ProcessBuilder) is abc
+        #else:
+        #    raise TypeError(
+        #        'builder has to be of type ProcessBuilder if no process_class is specified'
+        #    )
+        name = label + str(process_class).split('.')[-1].strip("'>") + '-nodes-' + bui_hash
+        print(name)
+
+        # check existence
+        full_import_path = str(data_dir) + '/' + name + '.tar.gz'
+        print(full_import_path)
+        cache_path = pathlib.Path(full_import_path)
+        if cache_path.exists():
+            cache_exists = True
+
+        if cache_exists:
+            # import data from previous run to use caching
+            load_cache(path_to_cache=full_import_path)
+
+        # now run process/workchain whatever
+        with enable_caching():  # should enable caching globally in this python interpreter
```

Comment:
Shouldn't it be possible to re-use `with_export_cache` here? That exports the calculations while here we explicitly export the parts connected to the node, but is this distinction really needed?

We could maybe change the `with_enable_caching` logic to only export calculations that were created _within_ the context manager (and their direct inputs / outputs). So, for example
```python
run(SomeCalculation, **<inputs>)
with with_export_cache(<...>):
    run(OtherCalculation, **<inputs>)
```
would export the `OtherCalculation`, but not `SomeCalculation`. That would also get around the need to always clean the DB before running the test.

Reply:
Currently run_with_cache (which is kind of a patch for run_get_node) exists only because I do not know how to explicitly do this and I want an automatic cache naming which changes if the test changes, otherwise debugging becomes harder. How do you know what is run in the `with` environment? One idea is query for creation time. Then on the other hand if you use as we currently do the default aiida export you get the full provenance, which means if your `SomeCalculation` and `OtherCalculation` are connected (which within a test they usually are) they will get exported both anyway.

Reply:
Yeah, I think it makes sense to just export what is created during the `with` statement (and its provenance). That can be either by creation time, or maybe by checking what exists already in the DB when entering the `with` statement - whatever works more reliably. 
I'm not sure if we might get into clock synchronization issues (between the DB and the python interpreter, maybe?) when using `ctime`.

Reply:
Comment: 'Spliting' the cache
works currently for both fixtures, since the export happens at the end of the with.
```
with with_export_cache(<...>):
    run(OtherCalculation, **<inputs>)
with with_export_cache(<...>):
    run(ThisWeCacheSeperateCalculation, **<inputs>)
```
So, if you know to extract what ran in the contextmanager let me know.

Reply:
I will have to do some testing to see if there is a good way to extract that...